File: Lab5/app/src/main/java/psy/fit/bstu/lab5/MainActivity.java
package psy.fit.bstu.lab5;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ListView;

import java.util.ArrayList;

import psy.fit.bstu.lab5.Listeners.OnContactClickListener;
import psy.fit.bstu.lab5.adapter.ContactAdapter;
import psy.fit.bstu.lab5.db.DatabaseContactController;

public class MainActivity extends AppCompatActivity {

    private ListView contactsList;
    private DatabaseContactController contactController;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        contactController = new DatabaseContactController(this);
        contactsList = findViewById(R.id.contactsListView);
        contactsList.setOnItemClickListener(new OnContactClickListener());
        loadContacts();
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        Log.d("main", "onrestart");

        loadContacts();

    }

    private void loadContacts() {
        ArrayList<Contact> contacts = (ArrayList<Contact>) contactController.read();
        if(contacts.size() < 1) {
            //show empty message
            return;
        }
        ContactAdapter adapter = new ContactAdapter(this, R.layout.contact_list, contacts);
        contactsList.setAdapter(adapter);
    }

    public void onCreateContact(View view) {
        Intent intent = new Intent(this, FormActivity.class);
        startActivity(intent);
    }
}

File: lab13/app/src/main/java/psy/fit/bstu/lab13/MainActivity.kt
package psy.fit.bstu.lab13

import IBaseModel
import Post
import Comments
import Todo
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import android.widget.RadioGroup
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import psy.fit.bstu.lab13.retrofit.CommonClient
import retrofit2.awaitResponse

class MainActivity : AppCompatActivity() {
    private lateinit var radioGroup: RadioGroup
    private lateinit var dataRecyclerView: RecyclerView
    private lateinit var progressBar: ProgressBar
    private lateinit var initText: TextView
    private val link = "https://jsonplaceholder.typicode.com"
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        this.initializeElements()
    }

    private fun initializeElements() {
        radioGroup = findViewById(R.id.radioApisGroup)
        dataRecyclerView = findViewById(R.id.dataRecycleView)
        progressBar = findViewById(R.id.progressbar)
        initText = findViewById(R.id.startLabel)

        this.radioGroup.setOnCheckedChangeListener { radioGroup, _ ->
            if (initText.visibility == View.VISIBLE) {
                initText.visibility = View.INVISIBLE
            }

            when (radioGroup.checkedRadioButtonId) {
                R.id.marvelApi -> {
                    this.renderHeroes()
                }
                R.id.postsApi -> {
                    this.renderPosts()
                }
                R.id.catsApi -> {
                    this.renderTodos()
                }
            }
        }
    }

    private fun renderHeroes() {
        progressBar.visibility = View.VISIBLE

        GlobalScope.launch(Dispatchers.IO) {
            val response = CommonClient(link).genericRetrofitService.getComments().awaitResponse()

            if (response.isSuccessful) {
                val data: List<IBaseModel> = response.body()!!

                withContext(Dispatchers.Main) {
                    render(data as List<Comments>)                }
            }
        }
    }

    private fun renderPosts() {
        progressBar.visibility = View.VISIBLE

        GlobalScope.launch(Dispatchers.IO) {
            val response = CommonClient(link).genericRetrofitService.getPosts().awaitResponse()

            if (response.isSuccessful) {
                val data: List<IBaseModel> = response.body()!!

                withContext(Dispatchers.Main) {
                    render(data as List<Post>)
                }
            }
        }
    }

    private fun renderTodos() {
        progressBar.visibility = View.VISIBLE

        GlobalScope.launch(Dispatchers.IO) {
            val response = CommonClient(link).genericRetrofitService.getTodos().awaitResponse()

            if (response.isSuccessful) {
                val data: List<IBaseModel> = response.body()!!

                withContext(Dispatchers.Main) {
                    render(data as List<Todo>)
                }
            }
        }
    }

    private fun render(data: List<IBaseModel>) {
        progressBar.visibility = View.GONE
        val layoutManager = LinearLayoutManager(this@MainActivity)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        dataRecyclerView.layoutManager = layoutManager
        dataRecyclerView.adapter = ViewAdapter(data)
    }
}